### PR TITLE
Avoid creating duplicate issues

### PR DESCRIPTION
**Pull request by SoapHia**
**Context**: To ensure that the latest dependency updates (on our base images) are added to the published images, we are planning to rebuild images for each release every fortnight. This requires a periodic scheduler to be added through Cloud Build and will run the same cloudbuild.yaml as release builds. 

This could produce duplicate Trivy scans for the same release, leading to duplicate GitHub issues. To avoid this, we must scan existing GitHub issues before creating new ones.

This PR refactors the code in the existing Trivy result parser to handle the following cases:

- If no duplicate issue exists: create a new issue if necessary
- If duplicate issue exists: overwrite the existing issue, or close it if the latest scan found no vulnerabilities

This PR also adds some unit tests for the parser's more complex functions. The unit tests have also been configured to run in a GitHub Action whenever the parser's code has been modified. 

Please find an example run of the new workflow [here](https://github.com/SoapHia/ts-bridge/runs/1666214440?check_suite_focus=true).

### CODE REVIEW

File: ci/parse_trivy_results.py [hunk 1]
```diff
@@ -131,54 +126,121 @@ def create_issue(target_name, num_vulnerabilities, severity_list, table):
...
+        repo, existing_issue, title, body)
+
+    # Print details to build log.
+    details = ("{} vulnerabilities of type: [{}] were found in image. "
+               "Please refer to issue #{} for details.").format(
+        num_vulnerabilities, ",".join(severity_list), issue_number)
+    print(details)
+
+    # Fail build if high or critical vulnerabilities are found.
+    if high_or_critical_exists(severity_list):
```
Comment (nerdinary) [inline ci/parse_trivy_results.py:205]: perhaps worth noting in this message why the build is being aborted?

File: ci/parse_trivy_results_tests.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        self.trivy_json_high = {"Vulnerabilities": [
+            {"Severity": "HIGH"}], "Target": "Test Target"}
+
+        self.trivy_json_mixed = {"Vulnerabilities": [
+            {"Severity": "LOW"}, {"Severity": "HIGH"}], "Target": "Test Target"}
+
+        self.mock_repo = mock.Mock()
+        self.mock_repo.create_issue.return_value = self.mock_issue_dup
+
+    def test_get_duplicate_issue_no_duplicates(self):
```
Comment (nerdinary) [inline ci/parse_trivy_results_tests.py:41]: Could you add a doc string describing what is being tested in each of these cases?

File: ci/parse_trivy_results.py [hunk 3]
```diff
@@ -131,54 +126,121 @@ def create_issue(target_name, num_vulnerabilities, severity_list, table):
...
+            print("WARNING: More than one vulnerability issue exists for this "
+                  "release. We will work with the most recently updated issue.")
+        return prev_issue[0]
+    else:
+        return None
 
-    return new_issue.number
+
+def process_vulnerabilities(repo, existing_issue, trivy_json, trivy_table):
+    """Update issue and build log based on non-empty vulnerabilities."""
```
Comment (nerdinary) [inline ci/parse_trivy_results.py:187]: Could you update the docstring to follow the format here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Include what the args are, and what the expected failures / returns are.

File: ci/parse_trivy_results.py [hunk 4]
```diff
@@ -131,54 +126,121 @@ def create_issue(target_name, num_vulnerabilities, severity_list, table):
...
+
+def create_or_update_issue(repo, existing_issue, title, body):
+    """Creates or updates a Github issue for the current release."""
 
     try:
-        new_issue = repo.create_issue(title=title, body=body)
+        if existing_issue:
+            existing_issue.edit(title=title, body=body)
+            return existing_issue.number
+        else:
```
Comment (nerdinary) [inline ci/parse_trivy_results.py:159]: you don't need an else here

**Review by nerdinary — COMMENTED** [review]
Generally looks good, nice work!

Could you run tests with coverage? (https://www.sealights.io/agile-testing/test-metrics/python-code-coverage/). This will tell you if the lines of code that we actually care about are being tested or not.

**Comment by Temikus** [comment]
My 0.02$ regarding test coverage: 
Those are two most popular services for that: https://about.codecov.io/ https://coveralls.io/

**Comment by SoapHia** [comment]
> Generally looks good, nice work!
> 
> Could you run tests with coverage? (https://www.sealights.io/agile-testing/test-metrics/python-code-coverage/). This will tell you if the lines of code that we actually care about are being tested or not.

Please find the results from coverage below:
```
Name                              Stmts   Miss  Cover   Missing
ci/parse_trivy_results.py           110     46    58%   56-71, 76-87, 102-113, 164-165, 236-241, 246-258, 262-269

**Review by nerdinary — APPROVED** [review]

**Review by Temikus — APPROVED** [review]

**Pull request merged**